# Analyzing Neighborhoods in Toronto

# Introduction

### Toronto is Capital of Canada. Being Major city attracts lot of people for work, tourism etc..Lets say if someone got a job in Toronto and want to relocate to the city and he is interested in choosing a neighborhood which has parks, outdoor activities, Gym etc. So we will analyze Neighborhoods of Toronto and try to find the neighborhoods which has above requirements. 

Before we get the data and start exploring it, let's download all the dependencies that we will need.

In [1]:
# Importing Libraries
import requests
!conda install -c anaconda beautifulsoup4 --yes
from bs4 import BeautifulSoup
from urllib.request import urlopen
import pandas as pd

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - beautifulsoup4


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    beautifulsoup4-4.9.1       |           py36_0         168 KB  anaconda
    ca-certificates-2020.1.1   |                0         132 KB  anaconda
    certifi-2020.6.20          |           py36_0         160 KB  anaconda
    openssl-1.1.1g             |       h7b6447c_0         3.8 MB  anaconda
    soupsieve-2.0.1            |             py_0          33 KB  anaconda
    ------------------------------------------------------------
                                           Total:         4.3 MB

The following NEW packages will be INSTALLED:

  beautifulsoup4     anaconda/linux-64::beautifulsoup4-4.9.1-py36_0
  soupsieve          anaconda/noarch::soupsieve-2.0.1-py_0

The following packag

## 1. Download and Explore Dataset

For the Toronto neighborhood data, a Wikipedia page exists, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M that has all the information we need to explore and cluster the neighborhoods in Toronto. We will scrape the Wikipedia page into a structured format using Pandas Dataframe


The data is available in table format with Postal code, Borough and Neighborhood as columns

In [2]:
# Using pandas to read the table into pandas dataframe
df=pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M', flavor='html5lib')
# Extract the first table in the page
df_neighborhood=df[0]  
df_neighborhood.head() 

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

The table consists of few cells with no Borough. Remove the cells in which Borough is Not assigned

In [3]:
df_neighborhood.drop(df_neighborhood[df_neighborhood['Borough']=='Not assigned'].index, inplace=True)
df_neighborhood.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

chech the shape of the dataframe

In [4]:
df_neighborhood.shape

(103, 3)

The dataframe has 103 postal codes. The table is grouped by postal codes. There are multiple Borough and Neighborhoods with same postal code. 
So we will go ahead and segment using postal codes.

Check for any duplicate postal codes or empty Borough cells

In [5]:
bool_series =df_neighborhood["Postal Code"].duplicated().any()
print(bool_series)

False


## 2. Download the Spatial coordinates for all postal codes

The spatial coordinates for all postal codes are available in 'Geospatial_Coordinates.csv'. Using the spatial coordinates we will explore the venues for all neighborhoods.


Load the Spatial coordinates in to pandas dataframe

In [6]:
df_coordinates=pd.read_csv('Geospatial_Coordinates.csv')
df_coordinates.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Merge the 2 dataframes grouped by Postal codes

In [7]:
df_toronto = pd.merge(df_neighborhood, df_coordinates, on='Postal Code', how='outer')
df_toronto.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [8]:
df_toronto.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103 entries, 0 to 102
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Postal Code   103 non-null    object 
 1   Borough       103 non-null    object 
 2   Neighborhood  103 non-null    object 
 3   Latitude      103 non-null    float64
 4   Longitude     103 non-null    float64
dtypes: float64(2), object(3)
memory usage: 4.8+ KB


## 3. Foursquare API

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.

#### Define Foursquare Credentials and Version

In [9]:
CLIENT_ID = 'KQYJCCZ32VQRTL1EJIASG4YTWEFP4R2TWNBIRFW2ITX2UMQA' # your Foursquare ID
CLIENT_SECRET = 'DPQOCMUFXJ1HXLECKTNMWJVBXSFSG0RNFW31SYUNRFQXMYME' # your Foursquare Secret
VERSION = '20200101' # Foursquare API version

print('Your credentails:')
#print('CLIENT_ID: ' + CLIENT_ID)
#print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:


#### Now, let's get the top 100 venues that are in Parkwoods within a radius of 500 meters.

In [10]:

LIMIT = 100 # limit of number of venues returned by Foursquare API


radius = 500 # define radius

In [11]:
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    43.753259, 
    -79.329656, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=KQYJCCZ32VQRTL1EJIASG4YTWEFP4R2TWNBIRFW2ITX2UMQA&client_secret=DPQOCMUFXJ1HXLECKTNMWJVBXSFSG0RNFW31SYUNRFQXMYME&v=20200101&ll=43.753259,-79.329656&radius=500&limit=100'

In [12]:
results = requests.get(url).json()["response"]['groups'][0]['items']
results

[{'reasons': {'count': 0,
   'items': [{'summary': 'This spot is popular',
     'type': 'general',
     'reasonName': 'globalInteractionReason'}]},
  'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
   'name': 'Brookbanks Park',
   'location': {'address': 'Toronto',
    'lat': 43.751976046055574,
    'lng': -79.33214044722958,
    'labeledLatLngs': [{'label': 'display',
      'lat': 43.751976046055574,
      'lng': -79.33214044722958}],
    'distance': 245,
    'cc': 'CA',
    'city': 'Toronto',
    'state': 'ON',
    'country': 'Canada',
    'formattedAddress': ['Toronto', 'Toronto ON', 'Canada']},
   'categories': [{'id': '4bf58dd8d48988d163941735',
     'name': 'Park',
     'pluralName': 'Parks',
     'shortName': 'Park',
     'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/parks_outdoors/park_',
      'suffix': '.png'},
     'primary': True}],
   'photos': {'count': 0, 'groups': []}},
  'referralId': 'e-0-4e8d9dcdd5fbbbb6b3003c7b-0'},
 {'reasons': {'count': 0,
   'items': [{'

We can see that all the information is in the *items* key.Lets extract the venues for all neifghborhoods

In [13]:
def getNearbyVenues(postal_code,neighborhood, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, neigh, lat, lng in zip(postal_code,neighborhood, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        try:
            results = requests.get(url).json()["response"]["groups"][0]['items']
        except:
            KeyError: 'groups'
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name,
            neigh,
            lat, 
            lng, 
            v['venue']['name'], 
            #v['venue']['location']['lat'], 
            #v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code',
                             'Neighborhood',
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *Toronto_venues*.

In [14]:
Toronto_venues = getNearbyVenues(postal_code=df_toronto['Postal Code'],
                                 neighborhood=df_toronto['Neighborhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )

In [15]:
Toronto_venues.head()

Postal Code      Neighborhood  Neighborhood Latitude  \
0         M3A         Parkwoods              43.753259   
1         M3A         Parkwoods              43.753259   
2         M4A  Victoria Village              43.725882   
3         M4A  Victoria Village              43.725882   
4         M4A  Victoria Village              43.725882   

   Neighborhood Longitude                   Venue         Venue Category  
0              -79.329656         Brookbanks Park                   Park  
1              -79.329656           Variety Store      Food & Drink Shop  
2              -79.315572  Victoria Village Arena           Hockey Arena  
3              -79.315572               Portugril  Portuguese Restaurant  
4              -79.315572             Tim Hortons            Coffee Shop

Lets check the shape of the Toronto_venues

In [16]:
Toronto_venues.shape

(2127, 6)

Lets check the total number of unique venue categories

In [17]:

print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 268 uniques categories.


In [18]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

Let's check how many venues were returned for each neighborhood

In [19]:
Toronto_groups=Toronto_venues.groupby('Neighborhood').count()

In [20]:
Toronto_groups

Postal Code  \
Neighborhood                                                      
Agincourt                                                     5   
Alderwood, Long Branch                                        6   
Bathurst Manor, Wilson Heights, Downsview North              21   
Bayview Village                                               4   
Bedford Park, Lawrence Manor East                            24   
Berczy Park                                                  58   
Birch Cliff, Cliffside West                                   4   
Brockton, Parkdale Village, Exhibition Place                 24   
Business reply mail Processing Centre, South Ce...           17   
CN Tower, King and Spadina, Railway Lands, Harb...           17   
Caledonia-Fairbanks                                           4   
Canada Post Gateway Processing Centre                        12   
Cedarbrae                                                     8   
Central Bay Street                                           65   
Christie                                                     16   
Church and Wellesley                                         77   
Clarks Corners, Tam O'Shanter, Sullivan                      13   
Cliffside, Cliffcrest, Scarborough Village West               2   
Commerce Court, Victoria Hotel                              100   
Davisville                                                   35   
Davisville North                                              8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn              5   
Don Mills                                                    27   
Dorset Park, Wexford Heights, Scarborough Town ...            6   
Downsview                                                    15   
Dufferin, Dovercourt Village                                 14   
East Toronto, Broadview North (Old East York)                 4   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...            9   
Fairview, Henry Farm, Oriole                                 67   
First Canadian Place, Underground city                      100   
Forest Hill North & West, Forest Hill Road Park               4   
Garden District, Ryerson                                    100   
Glencairn                                                     4   
Golden Mile, Clairlea, Oakridge                               8   
Guildwood, Morningside, West Hill                             8   
Harbourfront East, Union Station, Toronto Islands           100   
High Park, The Junction South                                24   
Hillcrest Village                                             4   
Humber Summit                                                 4   
Humberlea, Emery                                              1   
Humewood-Cedarvale                                            4   
India Bazaar, The Beaches West                               24   
Kennedy Park, Ionview, East Birchmount Park                   5   
Kensington Market, Chinatown, Grange Park                    59   
Kingsview Village, St. Phillips, Martin Grove G...            3   
Lawrence Manor, Lawrence Heights                             15   
Lawrence Park                                                 3   
Leaside                                                      33   
Little Portugal, Trinity                                     44   
Malvern, Rouge                                                1   
Milliken, Agincourt North, Steeles East, L'Amor...            2   
Mimico NW, The Queensway West, South of Bloor, ...           13   
Moore Park, Summerhill East                                   2   
New Toronto, Mimico South, Humber Bay Shores                 13   
North Park, Maple Leaf Park, Upwood Park                      4   
North Toronto West, Lawrence Park                            19   
Northwest, West Humber - Clairville                           3   
Northwood Park, York University                               6   
Old Mill South, King's Mill Park

In [21]:
Toronto_groups.shape

(94, 5)

## 3. Analyze Each Neighborhood

In [22]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe

Toronto_onehot.drop('Neighborhood', axis=1, inplace=True)
Toronto_onehot.insert(0,'Neighborhood',Toronto_venues['Neighborhood'])


Toronto_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0         Parkwoods                  0                  0        0   
1         Parkwoods                  0                  0        0   
2  Victoria Village                  0                  0        0   
3  Victoria Village                  0                  0        0   
4  Victoria Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0          0       0              0          0         0   
1          0       0              0          0         0   
2          0       0              0          0         0   
3          0       0              0          0         0   
4          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0         

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [23]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped.head()

Neighborhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                0.0      0.0                 0.0           0.0   
1                0.0      0.0                 0.0           0.0   
2                0.0      0.0                 0.0           0.0   
3                0.0      0.0                 0.0           0.0   
4                0.0      0.0                 0.0           0.0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0             0.0              0.0               0.0             0.000000   
1             0.0              0.0               0.0             0.000000   
2             0.0              0.0               0.0             0.000000   
3             0.0              0.0               0.0             0.000000   
4             0.0              0.0               0.0             0.041667   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0           0.0       0.0          0.0         0.0                  0.0   
1           0.0       0.0          0.0         0.0                  0.0   
2           0.0       0.0          0.0         0.0                  0.0   
3           0.0       0.0          0.0         0.0                  0.0   
4           0.0       0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0               0.0                 0.0            0.0        0.0         0.0   
1               0.0                 0.0            0.0        0.0         0.0   
2               0.0                 0.0            0.0        0.0         0.0   
3               0.0                 0.0            0.0        0.0         0.0   
4               0.0                 0.0            0.0        0.0         0.0   

   Bagel Shop  Bakery      Bank  Bar  Baseball Field  Baseball Stadium  \
0         0.0     0.0  0.000000  0.0             0.0               0.0   
1         0.0     0.0  0.000000  0.0             0.0               0.0   
2         0.0     0.0  0.095238  0.0             0.0               0.0   
3         0.0     0.0  0.250000  0.0             0.0               0.0   
4         0.0     0.0  0.000000  0.0             0.0               0.0   

   Basketball Court  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0               0.0                 0.0    0.0              0.0       0.0   
1               0.0                 0.0    0.0              0.0       0.0   
2               0.0                 0.0    0.0              0.0       0.0   
3               0.0                 0.0    0.0              0.0       0.0   
4               0.0                 0.0    0.0              0.0       0.0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0         0.0                 0.0        0.0     0.0            0.0   
1         0.0                 0.0        0.0     0.0            0.0   
2         0.0                 0.0        0.0     0.0            0.0   
3         0.0                 0.0        0.0     0.0            0.0   
4         0.0                 0.0        0.0     0.0            0.0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0        0.0       0.0                   0.0        0.200000      0.0   
1        0.0       0.0                   0.0        0.000000      0.0   
2        0.0       0.0                   0.0        0.000000      0.0   
3        0.0       0.0                   0.0        0.000000      0.0   
4        0.0       0.0                   0.0        0.041667      0.0   

   Bridal Shop  Bubbl

#### Let's confirm the new size

In [24]:
Toronto_grouped.shape

(94, 268)

First, let's write a function to sort the venues in descending order.

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [26]:
import numpy as np
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Café   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
0        Breakfast Spot          Skating Rink  Latin American Restaurant   
1        Sandwich Place           Coffee Shop                        Pub   
2                  Bank   Fried Chicken Joint              Shopping Mall   
3                  Bank   Japanese Restaurant         Chinese Restaurant   
4            Restaurant        Sandwich Place         Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Clothing Store           Yoga Studio      Doner Restaurant   
1                   Gym          Airport Gate        Farmers Market   
2           Bridal Shop                 Diner        Sandwich Place   
3    Dim Sum Restaurant                 Diner        Discount Store   
4              Pharmacy               Butcher             Juice Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Diner        Discount Store    Distribution Center  
1           Event Space  Ethiopian Restaurant      Electronics Store  
2         Deli / Bodega            Restaurant            Supermarket  
3   Distribution Center               Dog Run            Yoga Studio  
4       Thai Restaurant                   Pub             Hobby Shop

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [51]:
from sklearn.cluster import KMeans
# set number of clusters
kclusters = 8

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([3, 0, 3, 3, 3, 3, 3, 3, 3, 3], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [52]:
# add clustering labels
neighborhoods_venues_sorted.drop('Cluster Labels', axis=1, inplace=True)
neighborhoods_venues_sorted.insert(1, 'Cluster Labels', kmeans.labels_)
Toronto_merged=neighborhoods_venues_sorted

Toronto_merged = Toronto_merged.join(df_toronto.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head() 

Neighborhood  Cluster Labels  \
0                                        Agincourt               3   
1                           Alderwood, Long Branch               0   
2  Bathurst Manor, Wilson Heights, Downsview North               3   
3                                  Bayview Village               3   
4                Bedford Park, Lawrence Manor East               3   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                Lounge        Breakfast Spot          Skating Rink   
1           Pizza Place        Sandwich Place           Coffee Shop   
2           Coffee Shop                  Bank   Fried Chicken Joint   
3                  Café                  Bank   Japanese Restaurant   
4           Coffee Shop            Restaurant        Sandwich Place   

       4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0  Latin American Restaurant        Clothing Store           Yoga Studio   
1                        Pub                   Gym          Airport Gate   
2              Shopping Mall           Bridal Shop                 Diner   
3         Chinese Restaurant    Dim Sum Restaurant                 Diner   
4         Italian Restaurant              Pharmacy               Butcher   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0      Doner Restaurant                 Diner        Discount Store   
1        Farmers Market           Event Space  Ethiopian Restaurant   
2        Sandwich Place         Deli / Bodega            Restaurant   
3        Discount Store   Distribution Center               Dog Run   
4             Juice Bar       Thai Restaurant                   Pub   

  10th Most Common Venue Postal Code      Borough   Latitude  Longitude  
0    Distribution Center         M1S  Scarborough  43.794200 -79.262029  
1      Electronics Store         M8W    Etobicoke  43.602414 -79.543484  
2            Supermarket         M3H   North York  43.754328 -79.442259  
3            Yoga Studio         M2K   North York  43.786947 -79.385975  
4             Hobby Shop         M5M   North York  43.733283 -79.419750

## 5. Examine Clusters

#### Cluster 1

In [53]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0]

Neighborhood  Cluster Labels  \
1                        Alderwood, Long Branch               0   
42  Kennedy Park, Ionview, East Birchmount Park               0   
86                                    Westmount               0   
90                  Willowdale, Willowdale West               0   
91                                       Woburn               0   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
1            Pizza Place        Sandwich Place           Coffee Shop   
42        Discount Store           Coffee Shop    Chinese Restaurant   
86           Pizza Place           Coffee Shop        Sandwich Place   
90           Pizza Place           Coffee Shop        Discount Store   
91           Coffee Shop     Korean Restaurant           Yoga Studio   

   4th Most Common Venue      5th Most Common Venue 6th Most Common Venue  \
1                    Pub                        Gym          Airport Gate   
42      Department Store                Yoga Studio               Dog Run   
86    Chinese Restaurant  Middle Eastern Restaurant        Discount Store   
90                  Bank                   Pharmacy         Grocery Store   
91                 Diner             Discount Store   Distribution Center   

      7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1            Farmers Market           Event Space  Ethiopian Restaurant   
42       Dim Sum Restaurant                 Diner   Distribution Center   
86             Intersection      Doner Restaurant               Dog Run   
90  Comfort Food Restaurant            Comic Shop  Ethiopian Restaurant   
91                  Dog Run      Doner Restaurant            Donut Shop   

   10th Most Common Venue Postal Code      Borough   Latitude  Longitude  
1       Electronics Store         M8W    Etobicoke  43.602414 -79.543484  
42             Donut Shop         M1K  Scarborough  43.727929 -79.262029  
86       Department Store         M9P    Etobicoke  43.696319 -79.532242  
90      Electronics Store         M2R   North York  43.782736 -79.442259  
91              Drugstore         M1G  Scarborough  43.770992 -79.216917

#### Cluster 2

In [54]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1]

Neighborhood  Cluster Labels 1st Most Common Venue  \
49  Malvern, Rouge               1  Fast Food Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
49      Department Store    Falafel Restaurant           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
49  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  \
49             Drugstore            Donut Shop       Doner Restaurant   

   Postal Code      Borough   Latitude  Longitude  
49         M1B  Scarborough  43.806686 -79.194353

#### Cluster 3

In [55]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2]

Neighborhood  Cluster Labels  \
39                                   Humberlea, Emery               2   
58  Old Mill South, King's Mill Park, Sunnylea, Hu...               2   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
39        Baseball Field           Yoga Studio      Doner Restaurant   
58        Baseball Field           Yoga Studio      Doner Restaurant   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
39    Dim Sum Restaurant                 Diner        Discount Store   
58    Dim Sum Restaurant                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
39   Distribution Center               Dog Run            Donut Shop   
58   Distribution Center               Dog Run            Donut Shop   

   10th Most Common Venue Postal Code     Borough   Latitude  Longitude  
39   Fast Food Restaurant         M9M  North York  43.724766 -79.532242  
58   Fast Food Restaurant         M8Y   Etobicoke  43.636258 -79.498509

#### Cluster 4

In [56]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3]

Neighborhood  Cluster Labels  \
0                                           Agincourt               3   
2     Bathurst Manor, Wilson Heights, Downsview North               3   
3                                     Bayview Village               3   
4                   Bedford Park, Lawrence Manor East               3   
5                                         Berczy Park               3   
6                         Birch Cliff, Cliffside West               3   
7        Brockton, Parkdale Village, Exhibition Place               3   
8   Business reply mail Processing Centre, South C...               3   
9   CN Tower, King and Spadina, Railway Lands, Har...               3   
11              Canada Post Gateway Processing Centre               3   
12                                          Cedarbrae               3   
13                                 Central Bay Street               3   
14                                           Christie               3   
15                               Church and Wellesley               3   
16            Clarks Corners, Tam O'Shanter, Sullivan               3   
17    Cliffside, Cliffcrest, Scarborough Village West               3   
18                     Commerce Court, Victoria Hotel               3   
19                                         Davisville               3   
20                                   Davisville North               3   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn               3   
22                                          Don Mills               3   
22                                          Don Mills               3   
23  Dorset Park, Wexford Heights, Scarborough Town...               3   
24                                          Downsview               3   
24                                          Downsview               3   
24                                          Downsview               3   
24                                          Downsview               3   
25                       Dufferin, Dovercourt Village               3   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...               3   
28                       Fairview, Henry Farm, Oriole               3   
29             First Canadian Place, Underground city               3   
30    Forest Hill North & West, Forest Hill Road Park               3   
31                           Garden District, Ryerson               3   
32                                          Glencairn               3   
33                    Golden Mile, Clairlea, Oakridge               3   
34                  Guildwood, Morningside, West Hill               3   
35  Harbourfront East, Union Station, Toronto Islands               3   
36                      High Park, The Junction South               3   
37                                  Hillcrest Village               3   
38                                      Humber Summit               3   
40                                 Humewood-Cedarvale               3   
41                     India Bazaar, The Beaches West               3   
43          Kensington Market, Chinatown, Grange Park               3   
44  Kingsview Village, St. Phillips, Martin Grove ...               3   
45                   Lawrence Manor, Lawrence Heights               3   
46                                      Lawrence Park               3   
47                                            Leaside               3   
48                           Little Portugal, Trinity               3   
51  Mimico NW, The Queensway West, South of Bloor,...               3   
52                        Moore Park, Summerhill East               3   
53       New Toronto, Mimico South, Humber Bay Shores               3   
54           North Park, Maple Leaf Park, Upwood Park               3   
55                  North Toronto West, Lawrence Park               3   
56                Northwest, West Humber - Clairville               3   
57                    Nort

#### Cluster 5

In [57]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4]

Neighborhood  Cluster Labels  \
50  Milliken, Agincourt North, Steeles East, L'Amo...               4   
65                                           Rosedale               4   
70                                Scarborough Village               4   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
50            Playground                  Park           Yoga Studio   
65                  Park            Playground                 Trail   
70            Playground         Jewelry Store           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
50               Dog Run          Dessert Shop    Dim Sum Restaurant   
65           Yoga Studio               Dog Run          Dessert Shop   
70      Doner Restaurant    Dim Sum Restaurant                 Diner   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
50                 Diner        Discount Store   Distribution Center   
65    Dim Sum Restaurant                 Diner        Discount Store   
70        Discount Store   Distribution Center               Dog Run   

   10th Most Common Venue Postal Code           Borough   Latitude  Longitude  
50             Donut Shop         M1V       Scarborough  43.815252 -79.284577  
65    Distribution Center         M4W  Downtown Toronto  43.679563 -79.377529  
70             Donut Shop         M1J       Scarborough  43.744734 -79.239476

#### Cluster 6

In [58]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 5]

Neighborhood  Cluster Labels  \
81  The Kingsway, Montgomery Road, Old Mill North               5   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
81                 River                  Pool           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
81   Distribution Center      Department Store          Dessert Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
81    Dim Sum Restaurant                 Diner        Discount Store   

   10th Most Common Venue Postal Code    Borough   Latitude  Longitude  
81                Dog Run         M8X  Etobicoke  43.653654 -79.506944

#### Cluster 7

In [59]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 6]

Neighborhood  Cluster Labels  \
10                            Caledonia-Fairbanks               6   
26  East Toronto, Broadview North (Old East York)               6   
61                                      Parkwoods               6   
87                                         Weston               6   
93                                York Mills West               6   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
10                  Park         Women's Store                  Pool   
26                  Park           Pizza Place     Convenience Store   
61     Food & Drink Shop                  Park           Yoga Studio   
87     Convenience Store                  Park            Donut Shop   
93     Convenience Store                  Park            Donut Shop   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
10           Yoga Studio               Dog Run          Dessert Shop   
26          Dessert Shop    Dim Sum Restaurant                 Diner   
61    Dim Sum Restaurant                 Diner        Discount Store   
87    Dim Sum Restaurant                 Diner        Discount Store   
93    Dim Sum Restaurant                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
10    Dim Sum Restaurant                 Diner        Discount Store   
26        Discount Store   Distribution Center               Dog Run   
61   Distribution Center               Dog Run      Doner Restaurant   
87   Distribution Center               Dog Run      Doner Restaurant   
93   Distribution Center               Dog Run      Doner Restaurant   

   10th Most Common Venue Postal Code     Borough   Latitude  Longitude  
10    Distribution Center         M6E        York  43.689026 -79.453512  
26       Doner Restaurant         M4J   East York  43.685347 -79.338106  
61             Donut Shop         M3A  North York  43.753259 -79.329656  
87            Yoga Studio         M9N        York  43.706876 -79.518188  
93            Yoga Studio         M2P  North York  43.752758 -79.400049

#### Cluster 8

In [61]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 7]

Neighborhood  Cluster Labels  \
67  Rouge Hill, Port Union, Highland Creek               7   

   1st Most Common Venue       2nd Most Common Venue 3rd Most Common Venue  \
67                   Bar  Construction & Landscaping           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
67            Donut Shop                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
67   Distribution Center               Dog Run      Doner Restaurant   

   10th Most Common Venue Postal Code      Borough   Latitude  Longitude  
67              Drugstore         M1C  Scarborough  43.784535 -79.160497

In [62]:
!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim


Solving environment: done

# All requested packages already installed.



In [63]:
import matplotlib.cm as cm
import matplotlib.colors as colors
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


In [64]:
import folium
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 6. Conclusion

 The neighborhoods under cluster 1 (Red) has parks as the most popular venues. This suggests that these neighborhoods has lot of space covered with parks and would be ideal neighborhoods for anyone seeking to relocate.